hazelcast/src/main/java/com/hazelcast/map/impl/MapMergeRunnable.java

```diff
@@ -50,14 +50,19 @@ class MapMergeRunnable extends AbstractMergeRunnable<Object, Object, RecordStore
     protected void mergeStore(RecordStore store, BiConsumer<Integer, MapMergeTypes<Object, Object>> consumer) {
         int partitionId = store.getPartitionId();
 
-        store.forEach((BiConsumer<Data, Record>) (key, record) -> {
-            Data dataKey = toHeapData(key);
-            Data dataValue = toHeapData(record.getValue());
-            ExpiryMetadata expiryMetadata = store.getExpirySystem().getExpiryMetadata(dataKey);
-            consumer.accept(partitionId,
+        store.beforeOperation();
+        try {
+            store.forEach((BiConsumer<Data, Record>) (key, record) -> {
```

Comment:
Looks like that might be a long running operation, there should be some epoch refreshes, otherwise other threads might get stuck. Probably there are some refreshes somewhere inside, but I'm not sure about that.

Reply:
Sorry, haven't noticed the comment. I'll address it in the follow-up PR.

Reply:
No other thread should access associated hybrid-log at this point, record-store is detached from its global registry and it is in an effectively read only mode, we iterate over entries to create merge operations and in the end record-store is destroyed. 

@taburet @petrpleshachkov Please correct me if i'm wrong but seems to me no further changes is needed for this PR.

Reply:
@ahmetmircik a thread needs to be registered in the epoch, `store.beforeOperation()` does that for TS, but the epoch is global and shared across all the threads including partition threads, so there might be other threads registered in it, if the thread is not refreshing/bumping the epoch for a long time that will stop the epoch progress. As we discovered with Petr, the iterator itself is doing a periodic epoch refresh, so no changes needed, but we also discovered some problems with the iterator creation logic which Petr is trying to fix with the follow-up PR.

Reply:
@taburet just trying to understand better;
Here record-store is in read-only mode and only one thread, which is not partitioned, reads it and creates merge operations with this read data then this record-store is destroyed.

Could you please clarify, for this situation, the reasoning behind epoch refresh need?

Reply:
@ahmetmircik because the epoch instance is global and shared across all threads of a single HZ instance, other threads might be using the same epoch instance for other purposes (other hybrid logs, for instance). Is there a guarantee such that when `MapMergeRunnable` is invoked a thread running it is the only active thread (actively using the epoch) in the HZ instance?

Reply:
@taburet We can think MapMergeRunnable as a user thread which does operation calls to populate map then destroys itself with its state, this is the only mission of it. From this perspective it should have nothing to share with other threads.